Add a deployment command and publish pipeline

douglasjacobsen opened the pull request:
This merge adds a deployment command with publish and fetch as subcommands, and a publish deployment pipeline.

The deployment command can be used to publish (from) and fetch (into) deployments using a workspace definition.

douglasjacobsen committed:
Add a deployment command and push pipeline

This commit adds a deployment command with push and pull as
subcommands, and a push deployment pipeline.

The deployment command is useful to push a workspace into a fully
contained deployment to manage experiments in a remote location.

The push pipeline is used to create a deployment from a
ramble workspace. This encapsulates all of the artifacts the workspace
needed to perform its experiments. It captures the workspace's
configuration files, and creates a workspace specific object repository
that contains package, application, and modifier definitions that the
workspace needs to execute.

The push subcommand can create a tar of the deployment, and also push
the deployment to a remote location.

The fetch subcommand takes a path to a remote deployment, and overwrites
the contents of the current workspace with the deployment's.

rfbgo commented on lib/ramble/docs/workspace.rst:
```diff
@@ -184,6 +184,43 @@ application definition files, one can use:
+
+.. code-block:: console
+
+  $ ramble deployment publish -u <remote_url>
```
Awesome stuff! Should we also document the "fetch" side?

douglasjacobsen replied:
Yeah, and maybe (as we discussed offline) I'll change this to `push` and `pull`. To align with other tools that use these same commands.

rfbgo commented on lib/ramble/ramble/cmd/deployment.py:
```diff
@@ -0,0 +1,178 @@
+            index_data = sjson.load(f)
+
+        for file in index_data[publish_cls.index_namespace]:
+            src = args.deployment_path + '/' + file
```
We use this pattern here, but we should probably rely on something like `os.join` since paths with too many `/` break in gcs buckets

douglasjacobsen replied:
My main concern was in URLs that might not want to use the OS separator (i.e. windows). But, this would also break local deployments at the OS level if you were publishing to a different directory.

Do you have any thoughts on how to manage this aspect of it?

rfbgo replied:
So more like `urllib`'s `urljoin`?

douglasjacobsen replied:
Yes! But I also think it might be trickier than that... I wonder what urllib does if the root is `file://` Maybe I'll play with it some and see if that would be a better replacement than just injecting `/`.

rfbgo commented on lib/ramble/ramble/pipeline.py:
```diff
@@ -521,9 +512,136 @@ def _execute(self):
+        else:
+            f = open(aux_repo_conf, 'w+')
+            f.write('repos:\n')
+        f.write(f'- ../../{self.object_repo_name}\n')
```
There is some bug where if you do a `publish` for the same workspace we instead keep appending to files and end up with an invalid distribution. 

First publish:

```
./deployments/gke/configs/modifier_repos.yaml:- ../object_repo
./deployments/gke/configs/repos.yaml:- ../object_repo
```

After many publishes:
```
 $ grep -r "object_repo" .
./deployments/gke/configs/modifier_repos.yaml:- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo
./deployments/gke/configs/repos.yaml:- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo- ../object_repo
./deployments/gke/configs/auxiliary_software_files/repos.yaml:- ../../object_repo
./deployments/gke/configs/auxiliary_software_files/repos.yaml:- ../../object_repo
./deployments/gke/configs/auxiliary_software_files/repos.yaml:- ../../object_repo
./deployments/gke/configs/auxiliary_software_files/repos.yaml:- ../../object_repo
./deployments/gke/configs/auxiliary_software_files/repos.yaml:- ../../object_repo
```
I think we need to add some logic around cleaning out old deployments

douglasjacobsen replied:
Ah that's fun. OK, I can look at that.

rfbgo commented on lib/ramble/ramble/pipeline.py:
```diff
@@ -521,9 +512,136 @@ def _execute(self):
+                f = open(repos_config, 'a')
+            else:
+                f = open(repos_config, 'w+')
+                f.write('repos:\n')
```
should this be `repos` or `f{conf_file}` or similar? 

I think the modifier repo needs the string `modifier_repos` here (https://github.com/GoogleCloudPlatform/ramble/blob/develop/lib/ramble/ramble/schema/modifier_repos.py#L18)

douglasjacobsen replied:
Yeah, I think you're right. It should be a string instead of just `repos`.

douglasjacobsen commented:
@rfbgo  I think I addressed all of your comments. I'd like to squash these commits before this is merged, assuming you're happy with the changes though.

rfbgo commented:
> @rfbgo I think I addressed all of your comments. I'd like to squash these commits before this is merged, assuming you're happy with the changes though.

Looks good to me. Squash away and we can merge

douglasjacobsen commented:
@rfbgo All done!
